Fixed issues associated with the 'man' target

christian-krieg opened the pull request:
* The manpage app builder does not support warnings, so I added if-statements
  that detect when 'man' is the make target and don't issue warnings in this
  case

llucax commented on sphinxcontrib/aafig.py:
```diff
@@ -86,7 +86,7 @@ def run(self):
                     v = float(v) / 100.0
                 aafig_options[k] = v
-                del self.options[k]
+                #del self.options[k]
```
Why is this commented out?

christian-krieg replied:
Hey, sorry I totally overlooked this issue (it was a long day for me, but I wanted to contribute the fixes that worked for me). My Python interpreter threw an exception that the dictionary is altered during iteration; I couldn't find a reason why the option should be deleted after processing them, so I commented-out that line and I was able to use aafig :)

llucax replied:
Ah, I see. I think the options are removed because they are *consumed* by this extension, so they are not passed to other extensions that might not know them. That said I haven't used (or developed for) Sphinx for probably more than a decade now, so my knowledge about the internals are very limited.

One easy way to fix the *dictionary is altered during iteration* issue is to iterate over a copy of the keys:

```py
for k in tuple(self.options.keys()):
    v = self.options[k]
    ...  # Rest of the for loop
```
This way we keep the same logic (that seems to be working so far).

llucax commented on sphinxcontrib/aafig.py:
```diff
@@ -100,7 +100,8 @@ def render_aafig_images(app, doctree):
     merge_dict(format_map, DEFAULT_FORMATS)
     if aafigure is None:
-        app.builder.warn('aafigure module not installed, ASCII art images '
+         if app.builder.format != 'man':
```
The indentation seems to be inconsistent, it should be 4 spaces instead of 5.

christian-krieg replied:
You are right! Will fix that

llucax commented on sphinxcontrib/aafig.py:
```diff
@@ -100,7 +100,8 @@ def render_aafig_images(app, doctree):
     if aafigure is None:
-        app.builder.warn('aafigure module not installed, ASCII art images '
+         if app.builder.format != 'man':
+            app.builder.warn('aafigure module not installed, ASCII art images '
```
Could you emit a [warning.warn()](https://docs.python.org/3/library/warnings.html) if `app.builder.format == 'man'`? Otherwise users will be left clueless about why things are not working.

This applies to this line and all other lines with a `if app.builder.format != 'man':`.

christian-krieg replied:
Sure! Thanks a lot for that detailed code review! Totally appreciate that and I feel a bit ashamed now that I didn't put more effort in the first place to provide code with higher quality!

llucax replied:
No need to be ashamed :)
I can totally understand that people, under the uncertainty of how a project's maintainer could take a PR or if it will reply to it at all, would lean towards the *lazy* side :)

And again, thanks a lot for going through the effort to contribute this and even more for following up on it! :pray:

llucax reviewed (changes requested):
Hi, thanks for contributing this! I'm a bit worried that users will be left with no help if using the `man` builder and there is some issue with `aafig`, this is mainly why I'm requesting changes.

christian-krieg committed:
Fixed issues associated with the 'man' target

* The manpage app builder does not support warnings, so I added if-statements
  that detect when a builder does not support warnings. In this case, a
  warnings.warn() is emitted

christian-krieg commented:
Hey Guys,

Thanks a lot for the discussion! I changed the code to reflect your concerns. I added a ``warn()`` function that checks if a builder supports warnings, and emits a ``warnings.warn()`` if the builder does not implement a ``warn`` method.

Cheers,
Christian

llucax commented on sphinxcontrib/aafig.py:
```diff
@@ -77,7 +78,8 @@ def run(self):
         image_attrs = dict()
         own_options_keys = list(self.own_option_spec.keys()) + ['scale']
-        for (k, v) in self.options.items():
+        for (k, v) in tuple(self.options.keys()):
```
Did you tried this? I think this won't work, you need to only use one value to iterate.

```suggestion
        for k in tuple(self.options.keys()):
```

llucax commented on sphinxcontrib/aafig.py:
```diff
@@ -96,12 +98,22 @@ def run(self):
         return [image_node]
 
 
+def warn(app, message):
```
This is just a minor suggestion, you are welcome to leave it as is, but any reason not to take the `builder` instead of the `app` here?

llucax commented on sphinxcontrib/aafig.py:
```diff
@@ -96,12 +98,22 @@ def run(self):
-                'will be redered as literal text')
+        message = 'aafigure module not installed, ASCII art images will be' \
+            'rendered as literal text'
+        warn(app, message)
```
I would also just put the message in the call instead of creating a `message` variable (same for the other calls to `warn()`), but you are also welcome to leave it as is if you prefer or are just lazy to change it :)

llucax reviewed (changes requested):
Only the first comment would be a blocker to merge this, the other comments are just mild personal preferences so it's fine if you want to leave them as they are.

Thanks again for following up on this!
